Feature: optimize pan zoom

hunkim98 opened the pull request:
🚀 [ Resolves #88 ]
🚀 [ Resolves #90 ]

### Preview

<!-- Please leave screenshots since they help others understand what you have done -->

#### Without optimization (1 million pixels) 
https://github.com/hunkim98/dotting/assets/57612141/8ec6c6cf-ff88-4001-9cfa-60b70d307622

#### With optimization (1 million pixels) 
https://github.com/hunkim98/dotting/assets/57612141/9ecb0122-5e91-431a-a5f9-c425a87d7539




<br/>

### Changes

<!-- Name a title to your changes -->

## Optimize panning and zooming using `Offscreen Canvas`

- _[🎨Component] Use `Offscreen Canvas` for creating image bitmap when user has modified the grid canvas_

  - Previously, changing offset of zoom scale called our render function. This inevitably made our component run through a nested loop everytime there is a slight change in offset or zoom scale
  - I found it unnecessary to call a nested for loop everytime we do a render.
  - I came up with an idea to create an image bitmap once a user's interaction finishes. Then I used that image bitmap when the user pans or zooms. 
  - This highly upgraded the performance of panning or zooming for 1 million pixels!


<br/>

## Notes

<!-- Write a note if you have any -->
- There seems to be some lag at the moment user finishes interacting with the grid canvas. Maybe we should take a look into that issue
<br/>

## Next Up?
- With this optimization, I believe we could try building Photoshop from scratch?🙀 @Lee-Si-Yoon @lerrybe 
<!-- Write your next plans if you have any -->

hunkim98 commented on src/components/Canvas/DataLayer.tsx:
```diff
@@ -548,7 +555,106 @@ export default class DataLayer extends BaseLayer {
+      ? allRowKeys.length * squareLength * this.dpr
+      : allRowKeys.length * squareLength;
+
+    const canvas = new OffscreenCanvas(width, height);
```
The API seems to not cleanup some code since the API you provided is creating a canvas dom element and then changing it into a offscreencanvas. From what I know, `new OffscreenCavnas(width, height)` is the same thing but it is different in that it skips the creation of a canvas dom element.

Lee-Si-Yoon replied:
> is creating a canvas dom element and then changing it into a offscreencanvas

without creating a new dom element isn't it approachable by?

```js
const offscreenCanvas = this.element.transferControlToOffscreen();
```

hunkim98 replied:
Let me try to implement it. I am not sure if it will interfere with the original contents in the element.

hunkim98 replied:
I found that once you transfer control to offscreen, you cannot use it in the main thread. Thus I believe there is a need to create a new dom element if we want the worker thread to access the canvas and render it.

hunkim98 commented on src/components/Canvas/DataLayer.tsx:
```diff
@@ -29,6 +30,8 @@ export default class DataLayer extends BaseLayer {
   private layers: Array<DottingDataLayer>;
   private currentLayer: DottingDataLayer;
   private defaultPixelColor = DefaultPixelColor;
+  private capturedImageBitmap: ImageBitmap | null = null;
+  private capturedImageBitmapScale = 1;
```
From what I know, you cannot use image related APIs in an offscreen canvas. Actually I am not an expert of Offscreen canvas or workers, so I am not sure if how I implemented this feature is the correct way to do it. `ImageBitmap` is created in the offscreen canvas, and from what I know logic in offscreen canvas happens in the web worker. 

Maybe we have to add worker logic to this feature like the example below?
https://tech.kakao.com/2021/06/02/frontend-growth-09/

What is your opinion?

Lee-Si-Yoon replied:
For me it would be great to separate files and interest(what should have in canvas itself/ what should happen in worker API) and I assume creating offscreencanvas in the workerAPI and controlling sequence there(with postMessage & onMessage method) will optimize the overall flow.
For instance `lastRenderAllCall` can be replaced with onMessage, by giving permission of the offscreen logic to the workerAPI

hunkim98 replied:
Great I will change the code as you have suggested

hunkim98 commented:
@Lee-Si-Yoon I have attempted to implement Worker logic to our `Dotting`. However, there were many synchronization problems due to the library having multiple canvases. Worker logic is asynchronous and thus, it seemed that many of our canvases did not sync together. Thus I decided to discard the Worker logic and stick to our original version of simply rendering all by javascript logic. 

In the future, I believe implement worker logic will be impossible for `Dotting` due to its design of having multiple canvases 😥

hunkim98 commented:
> @Lee-Si-Yoon I have attempted to implement Worker logic to our `Dotting`. However, there were many synchronization problems due to the library having multiple canvases. Worker logic is asynchronous and thus, it seemed that many of our canvases did not sync together. Thus I decided to discard the Worker logic and stick to our original version of simply rendering all by javascript logic.
> 
> In the future, I believe implement worker logic will be impossible for `Dotting` due to its design of having multiple canvases 😥

However I do believe we can add a worker logic to the DataLayer render where it renders through all the rows and data. However, I believe that is another issue unrelated with optimizing pan zoom, so I will not add any code using worker logic in this PR

Lee-Si-Yoon commented:
I agree that implementing worker api is out of scope for this PR. Nice job let's merge this

+ discussed offline about `capturedImageBitmapScale` getting not initialized
+ just one last complain with responsibility, do we have to create fake canvas every time when `onInteractionEnded` is called

hunkim98 commented:
> I agree that implementing worker api is out of scope for this PR. Nice job let's merge this
> 
> * discussed offline about `capturedImageBitmapScale` getting not initialized
> * just one last complain with responsibility, do we have to create fake canvas every time when `onInteractionEnded` is called

I fixed the issue of `capturedImageBtimapScale` not being set correctly. Regarding creating fake canvas every time `onInteractionEnded` is called, I believe the logic would be more complex if we tried to separate parts for updating fake canvas and rendering pixel squares in them. Instead, I selectively rendered data layer and updated captured image only when data has been changed!

Lee-Si-Yoon reviewed (approved):
LGTM
